### PR TITLE
Fix creation of dataviews inferred with .NET types with sparse vectors

### LINKED ISSUE #586: Cursor creation fails on a dataview with sparse vectors
### System information

- **Windows 10.0.17134**:
- **.NET Version 2.1.301**: 

### Issue

- The creation of a cursor on a dataview with sparse vectors does not work anymore.
- Expected getter type is VBuffer<float>, produced getter type is VBuffer<VBuffer<float>>.

### Source code / logs

The following code fails.

class ExampleASparse
        {
            [VectorType(5)]
            public VBuffer<float> X;
        }
 
        [Fact]
        [TestCategory(Cat)]
        public void SparseDataView()
        {
            var inputs = new[] {
                new ExampleASparse() { X = new VBuffer<float> (5, 3, new float[] { 1, 10, 100 }, new int[] { 0, 2, 4 }) },
                new ExampleASparse() { X = new VBuffer<float> (5, 3, new float[] { 2, 3, 5 }, new int[] { 0, 1, 3 }) }
            };
            var host = new TlcEnvironment();
            var data = host.CreateStreamingDataView(inputs);
            VBuffer<float> value = new VBuffer<float>();
            int n = 0;
            using (var cur = data.GetRowCursor(i => true)) // it fails here
            {
                var getter = cur.GetGetter<VBuffer<float>>(0);
                while (cur.MoveNext())
                {
                    getter(ref value);
                    Assert.True(value.Count == 3);
                    ++n;
                }
            }
            Assert.True(n == 2);
            Done();
        }

### PR DESCRIPTION
**Pull request by xadupre**
Fixes #586. `DataViewConstructionUtils`'s methods to create dataviews over .NET types will now have correctly inferred getters in the case of sparse vectors.

### CODE REVIEW

File: test/Microsoft.ML.TestFramework/TestSparseDataView.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.ML.Runtime.Command;
```
Comment (TomFinley) [inline test/Microsoft.ML.TestFramework/TestSparseDataView.cs]: I'm fairly certain not all of these are used.

File: test/Microsoft.ML.TestFramework/TestSparseDataView.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            Done();
+        }
+
+        void GenericSparseDataView<T>(T[] v1, T[] v2)
+        {
+            var inputs = new[] {
+                new SparseExample<T>() { X = new VBuffer<T> (5, 3, v1, new int[] { 0, 2, 4 }) },
+                new SparseExample<T>() { X = new VBuffer<T> (5, 3, v2, new int[] { 0, 1, 3 }) }
+            };
+            var host = new TlcEnvironment();
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.TestFramework/TestSparseDataView.cs]: >var host = new TlcEnvironment(); [](start = 12, length = 32)

can you wrap it in using?

File: test/Microsoft.ML.TestFramework/TestSparseDataView.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        [TestCategory(Cat)]
+        public void SparseDataView()
+        {
+            GenericSparseDataView(new[] { 1f, 2f, 3f }, new[] { 1f, 10f, 100f });
+            GenericSparseDataView(new DvInt4[] { 1, 2, 3 }, new DvInt4[] { 1, 10, 100 });
+            GenericSparseDataView(new DvBool[] { true, true, true }, new DvBool[] { false, false, false });
+            GenericSparseDataView(new double[] { 1, 2, 3 }, new double[] { 1, 10, 100 });
+            GenericSparseDataView(new DvText[] { new DvText("a"), new DvText("b"), new DvText("c") },
+                                  new DvText[] { new DvText("aa"), new DvText("bb"), new DvText("cc") });
+            Done();
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.TestFramework/TestSparseDataView.cs]: >Done(); [](start = 12, length = 7)

You don't have to call Done if you don't do any basefiles comparison

File: test/Microsoft.ML.TestFramework/TestSparseDataView.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                {
+                    getter(ref value);
+                    Assert.True(value.Count == 3);
+                    ++n;
+                }
+            }
+            Assert.True(n == 2);
+        }
+
+        [Fact]
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.TestFramework/TestSparseDataView.cs]: >Fact [](start = 9, length = 4)

Thank you for fixing my mess!
Your code definitely do right thing, and it's nice to have tests, but I want to point one thing. Your tests never hit TypedCursor changes.
There is two conversions: one from some collection to IDataview through code in DataViewConstructionUltis, and from IDataVIew to Collection via TypedCursor and your tests cover only first case.

In order to invoke second case you need to do something like:
dataView.AsEnumerable<ClassWithArrays>(env, false).GetEnumerator() and enumerate over it.

Comment (xadupre) [reply test/Microsoft.ML.TestFramework/TestSparseDataView.cs]: I'll do it.

File: test/Microsoft.ML.TestFramework/TestSparseDataView.cs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Runtime.Tools;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Runtime.RunTests
+{
+    public sealed partial class TestSparseDataView : TestDataViewBase
```
[the remark this answers is not in the document]
Comment (xadupre) [reply test/Microsoft.ML.TestFramework/TestSparseDataView.cs]: Copy paste without thinking :)

**Comment by TomFinley** [comment]
Hi @xadupre thanks for addressing the comment! Clarifying note for the future, when I asked to modify the description, I meant the PR's description, not the PR's title. The description is still the little "we are excited to review your PR" built in text that we intend for people to write. I hope you don't mind, just going to change it so we can get this in since otherwise it looks good.

**Comment by xadupre** [comment]
Hi,

Sorry for that. I did not read this part, I just checked the checkboxes. I looked into the template used by scikit-learn and the text I forgot to replace was commented out (HTML comment) and here is the discussion about it (https://github.com/scikit-learn/scikit-learn/pull/6470). That might be a good idea for us too.

Xavier



Sent from Outlook<http://aka.ms/weboutlook>

________________________________
De : Tom Finley <notifications@github.com>
Envoyé : samedi 28 juillet 2018 16:59
À : dotnet/machinelearning
Cc : Xavier Dupre; Mention
Objet : Re: [dotnet/machinelearning] Fixes #586, the creation of getters on dataviews having sparse vectors (#587)


Hi @xadupre<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fxadupre&data=02%7C01%7Cxadupre%40microsoft.com%7Cbaa687d82f064276dfc808d5f49aa905%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636683867461378704&sdata=44hUPnHMRw855yU6XeDRnBVJqt5AbS8D47Xo3hpzsMU%3D&reserved=0> thanks for addressing the comment! Clarifying note for the future, when I asked to modify the description, I meant the PR's description, not the PR's title. The description is still the little "we are excited to review your PR" built in text that we intend for people to write. I hope you don't mind, just going to change it so we can get this in since otherwise it looks good.

**Pull request merged**